datasets/stream_spec.py

```diff
@@ -0,0 +1,192 @@
+import numpy as np
+import copy
+import collections
+from typing import Union
+
+
+class StreamSpecification:
+    def __init__(self, n_tasks: int, n_classes: int, random_seed: int = None,
+                 n_slots: Union[int, None] = None, n_drifts: Union[int, None] = None, sequential_drifts: bool = False) -> None:
+        """
+        Utility class that represents the class-task layout of the overall task stream.
+
+        Args:
+        n_tasks - number of tasks in the stream
+        n_slots - number of classes per task
+        """
+        assert n_tasks >= 2, 'need at least two tasks for continual learning'
+        assert n_slots is not None or n_drifts is not None or sequential_drifts, 'must specify drift type'
+
+        if n_slots is not None:
+            assert n_drifts == None and sequential_drifts == False, 'you must use n_slots, n_drifts or sequential_drifts arguments (cant use them together)'
+        if n_drifts is not None:
+            assert n_slots == None and sequential_drifts == False, 'you must use n_slots, n_drifts or sequential_drifts arguments (cant use them together)'
+        if sequential_drifts:
+            assert n_slots == None and n_drifts is None, 'you must use n_slots, n_drifts or sequential_drifts arguments (cant use them together)'
+
+        self.n_tasks = n_tasks
+        self.n_slots = n_slots
+        self.n_classes = n_classes
+        self.random_seed = random_seed
+        self.n_drifts = n_drifts
+        self.sequential_drifts = sequential_drifts
+
+        if n_drifts is not None:
+            self.task_classes = self.create_n_drifts()
+        elif sequential_drifts:
+            self.task_classes = self.create_sequential_drifts()
+        else:
+            self.task_classes = self.random_class_asigment()
+
+        self.current_task = 0
+
+    def create_n_drifts(self):
+        assert self.n_drifts < self.n_tasks
+
+        drift_duration = self.n_tasks // (self.n_drifts + 1)
+        drift_indexes = list()
+        for i in range(drift_duration, self.n_tasks, drift_duration):
+            if len(drift_indexes) == self.n_drifts:
+                break
+            drift_indexes.append(i)
+
+        if len(drift_indexes) < self.n_drifts:
+            drift_indexes.append(self.n_tasks-1)
+        print('Creating drifts at tasks: ', drift_indexes)
+
+        classes_per_task = self.n_classes // self.n_tasks
+        last_class = 0
+        last_drifted_class = -1
+        task_classes = list()
+        for t in range(self.n_tasks):
+            new_classes = list(range(last_class, last_class + classes_per_task))
+            if t in drift_indexes:
+                min_label = min(new_classes)
+                for c in range(last_drifted_class+1, min_label):
+                    new_classes.append(c)
+                last_drifted_class = min_label - 1
+            task_classes.append(sorted(new_classes))
+            last_class += classes_per_task
+
+        return task_classes
+
```

Comment:
You're right. We should have rethink how this code is working. I was kinda aware about the first issue, but have not come up with something better.
The notion about drifts affecting all past classes was from the drift literature - there in most cases it affects all classes, but I agree here it may be not reasonable to apply drift to all past classes.

Aside from that: thank you for really thorough code review.

Reply:
It is my pleasure. No problem!